### ISSUE TITLE
Normalization layer before attention and Mlp?

### ISSUE THREAD

**Issue by CDitzel** [lucidrains/vit-pytorch#56]
The code performs layer normalization prior to attention/mlp operations. However, the transformer architecture sketches clearly show the opposite order. Can someone clarify?

**Comment by lucidrains** [comment]
@CDitzel Hi Carsten! Astute observations :)

The very first Transformers uses what we now call post-normalization, where each branch is summed with the 
 residual and then layer normed. `norm(branch_fn(x) + x)`. This type of post-normalization transformer was hard to train, often diverging. People soon figured out that you need to warmup the learning rate gradually, and this alleviated the problem, though it did not completely solve the instability.

Then GPT came along, and they made a change to the architecture, where they moved this normalization to just before the branch, so it is `branch_fn(norm(x)) + x`. They found that using this trick led to much better stability, at the cost of a slightly decreased performance (though it seems at scale, it matters very little). A learning warmup is also no longer needed with this scheme, which has since been named pre-normalization. GPT3 also uses pre-normalization, so you see, at scale, it really no longer matters.

These days, more and more people are migrating to pre-normalization, with the exception of a few paper here and there https://arxiv.org/abs/2012.11747 . For a more in-depth comparison, you can read https://arxiv.org/pdf/2002.04745.pdf

**Comment by lucidrains** [comment]
@CDitzel Hi Carsten! Astute observations :)

The very first Transformer uses what we now call post-normalization, where each branch is summed with the
residual and then layer normed. norm(branch_fn(x) + x). This type of post-normalization transformer was hard to train, diverging often. This was alleviated with a scheduler with a learning rate warmup over a few thousand steps, although it did not completely prevent divergence.

Then GPT came along, and they made a change to the architecture, where they moved this normalization to just before the branch, so it is branch_fn(norm(x)) + x. They found that using this trick led to much better stability, at the cost of a slightly decreased performance (though it seems at scale, it matters very little). A learning warmup is also no longer needed with this scheme, which has since been named pre-normalization.

These days, more and more people are migrating to pre-normalization, with the exception of a few paper here and there https://arxiv.org/abs/2012.11747 . For a more in-depth comparison, you can read https://arxiv.org/pdf/2002.04745.pdf

**Comment by CDitzel** [comment]
thank you very much Phil for your detailed explanation. It is a pleasure following you and your course through the adventures of DL Best regards and a happy new year to come from Germany!
